src/Form/Control/Password.php

```diff
@@ -4,7 +4,46 @@
 
 namespace Atk4\Ui\Form\Control;
 
+use Atk4\Ui\Button;
+use Atk4\Ui\Js\JsExpression;
+
 class Password extends Line
 {
     public string $inputType = 'password';
+
+    /** By setting this variable to true you get an Eye-button on the right side which can toggle the visibility of the entered password */
+    public bool $eye;
+
+    #[\Override]
+    protected function init(): void
+    {
+        parent::init();
+
+        if ($this->eye) {
+            $button = new Button([
+                'class.grey' => true,
+                'iconRight' => 'eye slash',
+            ]);
+
+            $this->action = $button;
+
+            $button->on(
+                'click',
+                new JsExpression(
+                    <<<EOF
```

Comment:
```suggestion
                    <<<'EOF'
```

and pass all strings using JsExpression templation to prevent JS injections

Reply:
It's all good and you will have to guide me... :)

Anyway, yes,
1, getElementById works for the input field, how can I get the ID of the icon connected to the same field, it has a different ID.

I think I need more information on a lot of things, like Behat for exmple.

Reply:
> and pass all strings using JsExpression templation to prevent JS injections

see https://github.com/atk4/ui/blob/5.2.0/src/Js/JsExpression.php#L35, ie. `new JsExpression('let str = 'x' + [] + 'y';', ['arg'])`

> Also, a Behat (functional) test for testing the reveal functonality is needed.

see https://github.com/atk4/ui/tree/develop/tests-behat directory, create demo with password file and `password.feature` test which will check:
1. password is hidden
2. click reveal
3. password is shown
4. click reveal again
5. password must be hidden again

you can either setup a testing environment on your PC or use the CI to make it passing using trial and error

Reply:
Thanks for the info, I'm going to try and see what I can do about Behat, I've never used it before so....

Anyway, do you or anyone else have an idea on how to get the "ID" of the icon of the button?
In the code it looks like:
```
<i id="_b7947b56bcedc631d38d468ce0f74cda__voip_password_button_icon" class="  eye slash icon" style=""></i>
```

Reply:
ID - button has `->name` too, the same for inner icon probably

Reply:
@mvorisek  I'm aware of it, that's why I asked about it before your latest message.

 I checked the structure of the object but name and shortName are empty.
I really need help here to find the id of the button without using the "shortName" which is a part of the id.

Thanks in advance!


``` 
object(Atk4\Ui\Button)#2346 (27) {
  ["defaultTemplate"]=>
  string(11) "button.html"
  ["ui"]=>
  string(6) "button"
  ["icon"]=>
  NULL
  ["iconRight"]=>
  string(9) "eye slash"
  ["_jsActions":protected]=>
  array(0) {
  }
  ["model"]=>
  NULL
  ["region"]=>
  NULL
  ["class"]=>
  array(1) {
    [0]=>
    string(4) "grey"
  }
  ["style"]=>
  array(0) {
  }
  ["attr"]=>
  array(0) {
  }
  ["template"]=>
  NULL
  ["content"]=>
  NULL
  ["element"]=>
  string(3) "div"
  ["executorFactory":protected]=>
  NULL
  ["stickyArgs"]=>
  array(0) {
  }
  ["defaultName"]=>
  string(3) "atk"
  ["_addLater":protected]=>
  array(0) {
  }
  ["_rendered":protected]=>
  bool(false)
  ["_app":"Atk4\Ui\AbstractView":private]=>
  NULL
  ["maxNameLength"]=>
  int(60)
  ["uniqueNameHashes"]=>
  array(0) {
  }
  ["elements"]=>
  array(0) {
  }
  ["_elementNameCounts":"Atk4\Ui\AbstractView":private]=>
  array(0) {
  }
  ["_initialized":"Atk4\Ui\AbstractView":private]=>
  bool(false)
  ["name"]=>
  NULL
  ["_owner":"Atk4\Ui\AbstractView":private]=>
  NULL
  ["shortName"]=>
  NULL
}
```

Reply:
It is because the `$button` is not initialized, it needs to be part of the render tree first.

Reply:
Looking into https://github.com/atk4/ui/blob/5.2.0/src/Form/Control/Input.php#L167 I belive you can replace `new Button(...)` with `Button::addTo($this, ...)`. As such call is in `$this::init()`, `addTo(` will call the init on the button and after the call the button should have the ID present.

Reply:
I will try your find, I solved it by calling $this->renderView() for now but I think your find is more elegant if it works.

Reply:
Ok, I tried it and yes, it renders the button directly but gets a tag not found error on "content"... So, I'm sticking with the working solution for now.